Loosen last dim contiguity for sdpa constraint to include last dim 0,1

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139787

Previously we were checking for a last dim with stride == 1. When the size is <= 1 that also is sufficient because the stride is insignificant. Fix for https://github.com/pytorch/pytorch/issues/138317


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

eellison committed:
Loosen last dim contiguity for sdpa constraint to include last dim 0,1

[ghstack-poisoned]

eellison committed:
Update on "Loosen last dim contiguity for sdpa constraint to include last dim 0,1"


Previously we were checking for a last dim with stride == 1. When the size is <= 1 that also is sufficient because the stride is insignificant. Fix for https://github.com/pytorch/pytorch/issues/138317


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

shunting314 commented on torch/_inductor/lowering.py:
```diff
@@ -2349,9 +2349,12 @@ def is_aligned_realized_tensor(x):
-            return (
+            # if the last dim size is <= 1, stride doesnt matter
+            aligned_last_dim = (
                 V.graph.sizevars.size_hint(x.get_stride()[-1])
```
You missed the '== 1' part for stride?

eellison replied:
ahh good catch 😨

eellison committed:
Update on "Loosen last dim contiguity for sdpa constraint to include last dim 0,1"


Previously we were checking for a last dim with stride == 1. When the size is <= 1 that also is sufficient because the stride is insignificant. Fix for https://github.com/pytorch/pytorch/issues/138317


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting this, but I think it is failing on ROCm' -c nosignal

inductor/test_cuda_repro.py::CudaReproTests::test_scaled_dot_product_efficient_attention_backward [GH job link](https://github.com/pytorch/pytorch/actions/runs/11713218285/job/32626591565) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/060bee7f22a6ff5c14562713dc4bb6aa74923469)

Let me add `ciflow/rocm`

eellison commented:
cc @jeffdaily

eellison committed:
Update on "Loosen last dim contiguity for sdpa constraint to include last dim 0,1"


Previously we were checking for a last dim with stride == 1. When the size is <= 1 that also is sufficient because the stride is insignificant. Fix for https://github.com/pytorch/pytorch/issues/138317


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
